src/osd/OSD.h

```diff
@@ -805,12 +762,7 @@ class OSDService {
   void start_recovery_op(PG *pg, const hobject_t& soid);
   void finish_recovery_op(PG *pg, const hobject_t& soid, bool dequeue);
   bool is_recovery_active();
-  void release_reserved_pushes(uint64_t pushes) {
-    Mutex::Locker l(recovery_lock);
-    assert(recovery_ops_reserved >= pushes);
-    recovery_ops_reserved -= pushes;
-    _maybe_queue_recovery();
-  }
+  void release_reserved_pushes(uint64_t pushes);
```

Comment:
Yes, it looks like 3cc48278bf0ee5c9535d04b60a661f988c50063b introduced the bug by releasing the reserved items when requeueing.  The question is why we haven't noticed the problem on luminous.  I think the reason is simply that we never get into a state where we wake_pg_waiters() on a pg that has recovery items (the only things with reserved_pushes) queued, because the wake machinery is about pg instantiation (new pg, split pg, etc.).

https://github.com/ceph/ceph/pull/21229 should fix in this luminous, although I want to send it through several large rados runs before merging!

Reply:
Is this an issue that we/the nightlies have any hope of detecting?

That is, if we do too many releases, it just means extra recovery ops get allowed into the system. I don't think we have any way of noticing unless it happened so often that users complained about the performance change so loudly that we believed them.

Reply:
Yeah, it will assert at
```
  void release_reserved_pushes(uint64_t pushes) {
    Mutex::Locker l(recovery_lock);
    assert(recovery_ops_reserved >= pushes);
    recovery_ops_reserved -= pushes;
    _maybe_queue_recovery();
  }
```